en/02_Developer_Guides/02_Controllers/07_CMS_JSON_APIs.md

```diff
@@ -0,0 +1,154 @@
+
+## Return values and error handling
+
+All public endpoint methods must declare a return type of [`HTTPResponse`](api:SilverStripe\Control\HTTPResponse).
+
+All return values must utilise `jsonSuccess()` to create the `HTTPResponse` as this method is used to standardise JSON responses in the CMS.
+
+Do not throw exceptions in the controller, as this leads to a poor content-editor experience. Instead all non-success conditions must call `jsonError()`.
+
+### Using `jsonSuccess()`
+
+When using the optional `$data` parameter in `jsonSuccess()` to return JSON in the response body, do not add any "success metadata" around it, for example `['success' => true, 'data' => $data]`. Instead, solely rely on standard HTTP status codes to clearly indicate the success of the operation.
+
+For scenarios where no JSON data is returned in the response body upon success, use the status code 201 without the `$data` parameter i.e. `return $this->jsonSuccess(201);`. Alternatively, when the response includes JSON data, usually return a 200 status code i.e. `return $this->jsonSuccess(200, $data);`.
+
+### Using `jsonError()`
+
+The `return` keyword is omitted for `jsonError()` because internally it triggers an exception, subsequently caught and converted into an `HTTPResponse` object.
+
+Generally you should not include a message outlining the nature of the error when calling `jsonError()`, instead just include the appropriate HTTP status code for the error type, for example call `$this->jsonError(403)` if a permission check fails.
+
+If you do include a message, remember that error messages are only intended for developers so do not use the `_t()` function to make them translatable. Do not use any returned messages on the frontend for things like toast notifications, instead those messages should be added directly in JavaScript.
+
+[info]
+Despite the slightly convoluted JSON format returned by `jsonError()` with multiple nodes, its usage remains consistent with `FormSchema`. It's better to use this method for uniformity rather than introducing separate methods for `FormSchema` and non-FormSchema failures.
+[/info.]
+
+## CSRF token
+
+When performing non-view operations, include an `'X-SecurityID'` header in your JavaScript request, with its value set to `SecurityToken::getSecurityID()`.
+
+Access the token value in JavaScript by logging in by using `import Config from 'lib/Config';` followed by `Config.get('SecurityID');`
+
+Ensure the security of your endpoints by validating the token header using `SecurityToken::inst()->checkRequest($this->getRequest())` on relevant endpoints. If this check fails then call `$this->jsonError(400);` which is consistent with the 400 status code used in [`FormRequestHandler::httpSubmission()`](api:SilverStripe\Forms\FormRequestHandler::httpSubmission()) when the CSRF check fails when submitting data using `FormSchema`.
```

Comment:
````suggestion
Ensure the security of your endpoints by validating the security token on relevant endpoints.

```php
use SilverStripe\Security\SecurityToken;
// ...
if (!SecurityToken::inst()->checkRequest($this->getRequest())) {
    $this->jsonError(400);
}
```

[info]
The `400` HTTP status code used here is consistent with the code used when the CSRF check fails when submitting data using `FormSchema`.
[/info]
````

Again using a syntax-highlighted code snippet including use statements makes it clearer.

Reply:
Updated